common/cryptomsg.c

```diff
@@ -16,22 +21,88 @@
 #include <wire/wire_io.h>
 
 #if DISSECTOR
-static bool write_sk(char *sk)
+/* Map of key -> host.
+ * There are two entries for each host, one is the outgoing (sending key) 
+ * and the other is the incoming (receiving key).
+ *
+ * We store them by key so that we can lookup/replace them more easily.
+ */
+static STRMAP(const char *) keymap;
+
+/* Print keys to file in format: 
+ *  ->ip.add.re.ss:port sending_key
+ *  <-ip.add.re.ss:port receiving_key
+ *
+ *  i.e.
+ *    ->8.8.8.8:9735 0000000000000000000000000000000
+ *    <-8.8.8.8:9735 1111111111111111111111111111111
+ */
+static bool write_key(const char *key, const char *value, FILE *outfile)
+{
+	char buf[256];
+	snprintf(buf, sizeof buf, "%s %s\n", value, key);
+	if (fputs(buf, outfile) == -1) {
+		status_trace("Error printing keys.log file, exiting early");
+		return false;
+	}
+	return true;
+}
+
+static bool update_file(void)
 {
 	FILE *fp;
 	char *filename="keys.log";
 
-	fp = fopen(filename, "w+");
+	fp = fopen(filename, "w");
 	if (fp == NULL)
 		return false;
 
-	if (fputs(sk, fp) == -1)
-		return false;
+	strmap_iterate(&keymap, write_key, fp);
 
 	fclose(fp);
 	return true;
 }
-#endif
+
+void dissector_add_keys(const char *host, const char *sk, const char *rk)
+{
+	char host_out[256], host_in[256];
+	snprintf(host_out, sizeof host_out, "->%s", host);
+	snprintf(host_in, sizeof host_in, "<-%s", host);
+	strmap_add(&keymap, strdup(sk), strdup(host_in));
+	strmap_add(&keymap, strdup(rk), strdup(host_out));
+	update_file();
```

Comment:
@rustyrussell this works as intended.. if i connect to another peer, the logfile is updated as i'd expect